add customization of Discord button using invite URL from GMCP or Lua

atari2600tim opened the pull request:
<!-- Keep the title short & concise so anyone non-technical can understand it,
     the title appears in PTB changelogs -->
#### Brief overview of PR changes/additions
Changed the Discord button to use custom invite URL from game if available.  Added secondary drop-down item that goes to Mudlet Discord, have it be hidden if no custom invite.  Made GMCP handler save the URL.  Made Lua function for scripters to set the URL or clear it.  Updated the Discord logos.
#### Motivation for adding to Mudlet
Put that URL to use, if a game has implemented the GMCP message.
#### Other info (issues closed, discussion etc)
Closes #3115
#### Release post highlight
Discord button can go to custom invite URL, and new function to set it.
<!--
Use this space if you wish to write a short statement or example for inclusion
in the release post for the next release. 
-->

atari2600tim committed:
Discord button use custom invite URL from game

atari2600tim commented:
~~`ctelnet.cpp` has `if (mudlet::self()->mDiscord.libraryLoaded())` which determines if it should send `External.Discord.Hello` to the server.  `External.Discord.Info` is sent from server as a reply.  As a result, if you're trying this on Linux which does not support the Rich Presence, then you will also not get the `External.Discord.Info` containing the Discord invite URL.  The Lua function should work even on Linux though.~~ the cloud build is fine and includes the extras that weren't on my machine

vadi2 commented:
Linux does support Rich Presence... at least the official builds do. Are you running into this issue in your dev setup?

atari2600tim committed:
add comment regarding Discord privacy

atari2600tim commented:
> Linux does support Rich Presence... at least the official builds do. Are you running into this issue in your dev setup?

Oh, you're right, I see `lua gmcp` does have it on the automatically built one.

atari2600tim committed:
add hook into GMCP handler to enable visibility

atari2600tim committed:
name not needed here

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -10019,6 +10019,36 @@ int TLuaInterpreter::setDiscordApplicationID(lua_State* L)
+{
+    // The invite URL changes what the Discord button opens, it is not part of
+    // the rich presence, thus it does not have the api enabled check that all
+    // the other Discord functions have. This one is unrelated to privacy.
```
:thinking: That "not being related to privacy" and "not part of the rich presence" is worth pointing out...

SlySven commented on src/mudlet.cpp:
```diff
@@ -2353,10 +2370,26 @@ void mudlet::slot_irc()
     openWebPage(mMudletDiscordInvite);
 }
 
+void mudlet::toggleMudletDiscordVisible(bool vis){
```
Our coding style means this should be:
```suggestion
void mudlet::toggleMudletDiscordVisible(bool vis)
{
```

SlySven commented on src/mudlet.cpp:
```diff
@@ -563,7 +572,7 @@ mudlet::mudlet()
     connect(dactionForum, &QAction::triggered, this, &mudlet::slot_show_help_dialog_forum);
     connect(dactionIRC, &QAction::triggered, this, &mudlet::slot_irc);
-    connect(dactionDiscord, &QAction::triggered, this, &mudlet::slot_discord);
+    connect(dactionDiscord, &QAction::triggered, this, &mudlet::slot_mudlet_discord);
```
:-1: This changes the action of the existing main menu item - instead you should have an additional menu item (e.g. `dactionMudletDiscord`) that you show/hide as appropriate and connected to this new slot leaving the original menu menu item connected to the `slot_discord` slot method...

vadi2 replied:
Though I say different in https://github.com/Mudlet/Mudlet/issues/3115 (back 2 years ago), I agree with this - adding a separate button for the game-specific discord would be good. Seems like a shame to lose the access to the Mudlet discord which is super helpful for people.

atari2600tim replied:
With the edit a few hours ago, now the menu bar is similar to the button.  I was hesitant to put the custom game one under Help, which is why I had the Help menu show only the Mudlet-specific one.  But still I do like the idea of having the game one be available in the menu bar.  Also there is already IRC (which can be pointed at any server and channel) under Help, so I guess that's an alright place to put it.  Maybe in the future it and Help->IRC could be rearranged some without expanding scope here.

SlySven commented on src/mudlet.cpp:
```diff
@@ -1413,6 +1422,14 @@ void mudlet::slot_tab_changed(int tabID)
+    // If game has custom invite then make secondary Discord option visible.
+    bool isVis = mpActionMudletDiscord->isVisible();
+    if ( mpCurrentActiveHost->getDiscordInviteURL().isEmpty() && isVis ) {
+        mpActionMudletDiscord->setVisible(false);
```
As mentioned in another comment this `if` should also show/hide a new menu item (I previously suggested `dactionMudletDiscord`) as well...

SlySven commented on src/mudlet.cpp:
```diff
@@ -1413,6 +1422,14 @@ void mudlet::slot_tab_changed(int tabID)
 
+    // If game has custom invite then make secondary Discord option visible.
+    bool isVis = mpActionMudletDiscord->isVisible();
+    if ( mpCurrentActiveHost->getDiscordInviteURL().isEmpty() && isVis ) {
```
i suggest doing the computationally simplest test first! Also our coding style does not put spaces between the parenthesis of the `if` and what they contain - so this line would become:
```suggestion
    if (isVis && mpCurrentActiveHost->getDiscordInviteURL().isEmpty()) {
```

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -10019,6 +10019,36 @@ int TLuaInterpreter::setDiscordApplicationID(lua_State* L)
+
+    if (!lua_gettop(L)) {
+        host.setDiscordInviteURL(QString());
+        pMudlet->toggleMudletDiscordVisible(false);
```
:thinking: I think the calls to `mudlet::toggleMudletDiscordVisible(const bool)` might be being misused in this function - there is no guaranty that this profile is the active one and thus the one whose settings determine whether the main toolbar button (and the main menu item I talk about elsewhere in my review) should be "altered" to meet there being a custom Discord invite URL for this profile!

SlySven commented on src/mudlet.cpp:
```diff
@@ -218,6 +218,7 @@ mudlet::mudlet()
 , mpActionFullScreenView(nullptr)
 , mpActionHelp(nullptr)
 , mpActionDiscord(nullptr)
+, mpActionMudletDiscord(nullptr)
```
:mage_man: As it happens we are moving away from initialisation lists in class' constructors and this now ought to be done in the declaration (in the class header file) - furthermore as these are all `QPointer<T>` types they do not need an initialisation anyway - the default construction process will already ensure they are `nullptr`...

SlySven commented on src/mudlet.cpp:
```diff
@@ -407,11 +408,18 @@ mudlet::mudlet()
     mpActionDiscord->setObjectName(QStringLiteral("openDiscord"));
 
+    mpActionMudletDiscord = new QAction(tr("Mudlet Discord"), this);
+    mpActionMudletDiscord->setIcon(QIcon(QStringLiteral(":/icons/Discord-Logo-Color.png")));
```
I am not sure it is appropriate to use the same icon for two different things - we might want to consider adding a "badge" of, perhaps, the *Mudlet* arch icon for one of them - assuming that we are allowed to do that and do not fall foul of [Discord's "Branding" requirements](https://discord.com/branding)...

atari2600tim committed:
Update src/mudlet.cpp

missed one where { shouldn't be on line with if statement

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

atari2600tim committed:
Update src/mudlet.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
>> ~~`ctelnet.cpp` has `if (mudlet::self()->mDiscord.libraryLoaded())` which determines if it should send `External.Discord.Hello` to the server. `External.Discord.Info` is sent from server as a reply. As a result, if you're trying this on Linux which does not support the Rich Presence, then you will also not get the `External.Discord.Info` containing the Discord invite URL. The Lua function should work even on Linux though.~~ the cloud build is fine and includes the extras that weren't on my machine
>
>Linux does support Rich Presence... at least the official builds do. Are you running into this issue in your dev setup?

Of course FreeBSD does not support rich presence (unless one of our more skilful contributor did manage to get his FreeBSD boxen to run the Linux application using FreeBSD's "Linuxulator" Linux binary compatibility layer) and neither will 32-bit Linux boxes (older Raspberry Pis for example) so they are not going to get much out of that - but even then they can still access Discord via a web-browser...

atari2600tim committed:
make the help menu have both (instead of just the Mudlet) similar to how button has both

atari2600tim committed:
utilize function and simplify the enable part

I wrote this part before I wrote that function which did it

atari2600tim committed:
oops, I meant to make other menu item

atari2600tim committed:
and this one

atari2600tim committed:
check if current profile or not

atari2600tim committed:
Merge branch 'Mudlet:development' into tim-discord-URL

atari2600tim committed:
add combined icon

atari2600tim committed:
forgot the main one

atari2600tim committed:
Update mudlet_discord.png

SlySven commented on src/Host.cpp:
```diff
@@ -2463,6 +2465,17 @@ const QString& Host::getDiscordApplicationID()
+    writeProfileData(QStringLiteral("discordInviteURL"), s);
+}
+
+const QString& Host::getDiscordInviteURL()
```
At least consider this suggestion (and a matching one for the declaration in the `Host.h` file):
```suggestion
const QString& Host::getDiscordInviteURL() const
```
however I think it is even better to move this simple *getter* to the header file entirely - there are plenty of others of the same generic format already there - and they too should {but apart from a stopwatch one,  **don't**, :man_shrugging:) have the `const` *after* the method name {to flag that they are not capable of modifying the class's own members}!

atari2600tim committed:
simplify getter

SlySven reviewed (commented):
I am sorry but the `./src/icons/mudlet_discord.png` icon is not going to be suitable as is - I think you **have** to place it on a white, rectangular background that completely hides that portion of the mudlet arch - see the "Spacing" part of https://discord.com/branding ...! I am not exactly sure of how big the white background will have to be - but reading through the `Discord Brand Guidelines.pdf` inside the **[Discord media kit](https://www.dropbox.com/sh/nabhhaq7kt59exr/AAB7U3f2pW-Jmvdul0yy7o-ia?dl=1)** might help...

atari2600tim committed:
update art

Apply new logo that Discord switched to in May 2021 in the About screen and connection screen, replace my programmer art for the menu button.

atari2600tim committed:
Update TLuaInterpreter.cpp

atari2600tim committed:
Merge branch 'Mudlet:development' into tim-discord-URL

atari2600tim committed:
Merge branch 'Mudlet:development' into tim-discord-URL

atari2600tim committed:
token change to trigger github action

atari2600tim committed:
revert token change to trigger github action

atari2600tim committed:
token change to trigger github action

atari2600tim committed:
token change to trigger github action

vadi2 commented:
Just in case this takes a while to review, I've labelled this PR as accepted for the purposes of hacktoberfest so it sounds towards your shirt 👍

vadi2 commented on src/lua-function-list.json:
```diff
@@ -377,6 +377,7 @@
     "setDiscordDetail": "setDiscordDetail(text)",
     "setDiscordElapsedStartTime": "setDiscordElapsedStartTime(time)",
     "setDiscordGame": "setDiscordGame(text)",
+    "setDiscordInviteURL": "setDiscordInviteURL(url)",
```
This is autogenerated, don't worry about populating it

Kebap replied:
It is, if there is actually setDiscordInviteURL documented in wiki, but that is still missing, so I put a label here to work it out

vadi2 commented on src/mudlet.cpp:
```diff
@@ -2353,10 +2368,30 @@ void mudlet::slot_irc()
+        mpActionMudletDiscord->setVisible(vis);
+    }
+    if ( dactionDiscord->isVisible() != vis ) {
+        dactionDiscord->setVisible(vis);
```
```suggestion
    if (mpActionMudletDiscord->isVisible() != vis) {
        mpActionMudletDiscord->setVisible(vis);
    }
    if (dactionDiscord->isVisible() != vis) {
        dactionDiscord->setVisible(vis);
```

vadi2 reviewed (commented):
If you try to open Discord on a new profile before you're connected to anything, you'll get rick rolled 🤣 

I'm not sure on the UX here of silently swapping Mudlet's discord button with the game one now that I look at it. I know I thought it was a good idea a few years ago - doesn't seem like such a good one now. You can't tell which Discord you'll end up in when you press the button.

![image](https://user-images.githubusercontent.com/110988/137577547-a69bf031-7f70-42d3-9cb2-d887b27de943.png)


Another problem is that people don't discover the Mudlet discord so easily anymore - very few will bother to click the 'down'  button... and a lot of people find the Mudlet discord helpful, so that would be a pity. What do others think?

@Mudlet/ui-review

atari2600tim committed:
Update src/mudlet.cpp

Co-authored-by: Vadim Peretokin <vperetokin@hey.com>

SlySven reviewed (dismissed):
Sorry to put a *Request changes* on this - but my comment about needing a (probably) white background behind the Discord logo (on top of the Main Mudlet arch icon) so that we respect the Discord branding requirements in the `./src/icons/mudlet_discord.png` icon file still stands.

:thinking: Actually, white is not necessarily going to work in all situations, instead it'll have to be a transparent "hole" in the Mudlet arch icon so that the toolbar/menu background shows through...

atari2600tim commented:
> Another problem is that people don't discover the Mudlet discord so easily anymore - very few will bother to click the 'down' button... and a lot of people find the Mudlet discord helpful, so that would be a pity. What do others think?

I do like the concept of being able to tell someone on reddit or whatever to just click the obvious button and reach the game client chat, and this setup would lose some of that simplicity.

Some setups I'm not necessarily recommending but listing for sake of discussion: 

1. Current release: Discord button goes to Mudlet chat, invite from game goes to `/dev/null`
2. Current PR: When there is no custom invite, Discord button goes to Mudlet invite.  When there is a custom invite, Discord button now goes to custom invite, secondary drop-down option appears with alternate Discord-on-Mudlet icon to go to Mudlet invite.  Help menu does similar.
3. Something like 2 but switch them around... main button continues to go to Mudlet chat, but new drop-down option appears for custom invite.  It would have a new icon, the logo of Discord combined with something to do with games or MUDs.  Then game creators feel like theirs is hidden, but at least available.  Maybe a message could balance that out, like `[ INFO ] - This game offers an invite on Discord`.
4. Expand toolbar.  It would appear next to the existing Discord button or at the end and be called 'Game Discord' or 'Game chat' or something similar.  New alternate game-related icon (to be determined).
5. Combine 2 & 4... have the invite change behavior of plain Discord logo icon, and present the Discord-on-Mudlet icon but have it be alongside instead of 'hidden' in drop-down.  Slightly more crowded but everything visible, and the Mudlet logo is associated with Mudlet.

SlySven commented:
~~Shi~~ *Sugar* - that is not as hot as I had hoped, though it **does** respect the spacing that is mandated - this is a screen shot in G.I.M.P:
![image](https://user-images.githubusercontent.com/6163092/137597035-2980cad4-6c61-412c-b12e-8e1cb7c760ad.png)

This is the (256x256px) file as something you can download
![mudlet_with_discord_branding](https://user-images.githubusercontent.com/6163092/137597207-fbc55cc3-7760-46d9-a79e-f08a6a19bd28.png):

atari2600tim commented:
> 🤔 Actually, white is not necessarily going to work in all situations, instead it'll have to be a transparent "hole" in the Mudlet arch icon so that the toolbar/menu background shows through...

How does this look?  Clyde is even smaller but still identifiable I think, with it being near the main one. 
https://github.com/Mudlet/Mudlet/pull/5490/commits/9cd96f79334b62f6ac3d7997a7c206b243f923a4?short_path=3406d75#diff-3406d75af4c873339cc2fb650f40c438276357e8039b094f2585f102ae596184

edit: well that's a strange way of displaying the link, but it works I guess

SlySven commented:
I do not think the white-clyde-on-a-blue circle (screen shot of the above):
![image](https://user-images.githubusercontent.com/6163092/137597695-5d89dd0a-d6b0-41cd-8d47-b1b58222e7c9.png)
meets the spacing around Clyde that the guidelines stipulate (and which is demonstrated in the "Spacing" section of https://discord.com/branding ) - as you can see from my attempt the space around him is a bit more than one would expect.

I suppose we could contact brand@discordapp.com and get the word from the horse's mouth?

atari2600tim commented:
That circle one has the ratio and position of their own favicon.ico and likely the app, and has a single `o` of distance from the edge.  Their branding guidelines have a single `o` of distance between Clyde and the word, but they have 2 `o` of distance around the icon alone.  I'd agree that their web icon and app icon are a special case though because they expect some other space around it in a typical web browser or app listing, so they probably do indeed prefer not have just one `o` and then overlap with things perhaps even if the 2 things are paired together.  Mudlet paired together with Discord logo is something different from Discord logo being paired with the word Discord.

I made this rounded rectangle myself with 2`o` of distance from the edge, made it slightly larger than the previous circle so that Clyde wouldn't be shrunken too bad.
![two_circles](https://user-images.githubusercontent.com/29287358/137640221-7a934099-d2ef-4ac8-911b-1d5f52767b28.png)
Here's how it comes out for me, but likely better on 4k screen.  Even on my cheap screen though, context is enough to figure out what it is.
![image](https://user-images.githubusercontent.com/29287358/137640181-5bbd05a1-bb18-4008-8790-b18becbbdb4e.png)
Another alternative would be to just put no art or identical icon on that menu option.  Then it'd be like the `Connect` or `Packages (exp.)` menus with either a gap or repeats.

atari2600tim committed:
switch favicon to `oo` spacing

atari2600tim committed:
Merge branch 'Mudlet:development' into tim-discord-URL

vadi2 commented:
I like (5). @Kebap @keneanung @SlySven could you check quick so we make it into the feature freeze in two days?

atari2600tim committed:
try different layout on bar

Move it to separate item on bar, also I noticed that it didn't hide itself when opening a second tab, only upon switching, so fixed that too.

atari2600tim committed:
Update mudlet.cpp

improve tooltip

atari2600tim commented:
> I like (5).

I moved it there, hopefully didn't mess anything up.  Looks easier to recognize compared to the quarter of an already small icon deal from before.  Also caught where it would stay enabled when opening second profile.
![image](https://user-images.githubusercontent.com/29287358/138339127-f6f3b56c-d034-4420-9392-6483bc8cd184.png)

keneanung commented:
I like this version more as well. Though I'd move the arrow to the Mudlet Discord button, since the IRC goes to ours as well.

atari2600tim committed:
Merge branch 'Mudlet:development' into tim-discord-URL

atari2600tim committed:
Update mudlet.cpp

atari2600tim committed:
Update mudlet.cpp

simplify if statement

vadi2 commented:
What @keneanung said. I also had an idea - why not make the first button show the game name? That way you have the Discord icon + "Tim's game" underneath it, for example - and it is exactly intuitive what it will open.

I made a PR to your fork that does this: https://github.com/atari2600tim/Mudlet/pull/79

gesslar commented:
Will the game's discord icon pull the game's art for the button when it has that information?

vadi2 commented:
I don't think there's a way to transmit that in the protocol yet, but that is a good idea 👍

vadi2 committed:
Improve: Show game name on button if available (#79)

* Show game name on button if available

* Trigger DangerJS

atari2600tim committed:
change Lua function

change setDiscordInviteURL(url) to setDiscordGameUrl(url,name), combine the visibility code in with the name display code

atari2600tim committed:
Merge branch 'Mudlet:development' into tim-discord-URL

atari2600tim committed:
Update TLuaInterpreter.cpp

atari2600tim committed:
fix the url without name part I messed up

vadi2 commented:
Something is going wrong somewhere... the button isn't updating text. Hmm.

atari2600tim commented:
> Something is going wrong somewhere... the button isn't updating text. Hmm.

On b77f8d2 ?

vadi2 commented:
yeah, but it could have also been me because at the time I couldn't get stick to give the invite url

atari2600tim commented:
The specs say you don't get the message with the invite URL unless you have the presence checkbox turned on, because it is a response to the client saying Hello.  Is there any drawback to sending the Hello every time?

vadi2 commented:
Yeah, it's supposed to signify that the client is interested in Discord - so if it actually wasn't, that'd be a lie

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -10019,6 +10019,42 @@ int TLuaInterpreter::setDiscordApplicationID(lua_State* L)
+    }
+    QString inputText = getVerifiedString(L, __func__, 1, "url").trimmed();
+    host.setDiscordInviteURL(inputText.isEmpty() ? QString() : inputText);
+    if (args > 1){
```
```suggestion
    if (args > 1) {
```

atari2600tim committed:
Update src/TLuaInterpreter.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/TLuaInterpreter.cpp:
```diff
@@ -10019,6 +10019,42 @@ int TLuaInterpreter::setDiscordApplicationID(lua_State* L)
+    host.setDiscordInviteURL(inputText.isEmpty() ? QString() : inputText);
+    if (args > 1){
+        inputText = getVerifiedString(L, __func__, 2, "game name").trimmed();
+        host.setDiscordGameName(inputText.isEmpty() ? QString() : inputText);
```
Surely the ternary op is redundant? If `inputText.isEmpty()` that is functionally (except for `QString::isNull()` being able to tell the difference) the same as `QString()`? If we are not relying on that small detail then it can be simplified to:
```suggestion
        host.setDiscordGameName(inputText);
```

SlySven commented on src/mudlet.cpp:
```diff
@@ -2365,10 +2376,41 @@ void mudlet::slot_irc()
+void mudlet::updateDiscordNamedIcon()
+{
+    Host* pHost = getActiveHost();
+    if (!pHost) {
```
:thinking: If there *isn't* a profile currently active do we not want to return to the default, open the *Mudlet* Discord Guild in that case - in which case bailing out here isn't going to be enough - instead we should proceed on the basis that `gameName` should be a null/empty `QString` and `hasCustom` should be `false`?

atari2600tim replied:
If there isn't a profile active, then it would have the initial behavior of using the regular Discord button for the Mudlet Discord.  So don't want to do anything in that case.  Although... would anything be calling updateDiscordNamedIcon in that case?  It's called by GMCP handling, the Lua function, the tab change, and addConsoleForNewHost (late enough that it has been set), so I think they're all situations where there's an active host.  I think it'd be okay to just take that if statement out.

atari2600tim committed:
Update src/TLuaInterpreter.cpp

Co-authored-by: Stephen Lyons <slysven@virginmedia.com>

vadi2 commented:
I see what you mean with https://github.com/Mudlet/Mudlet/pull/5490#issuecomment-949957251 now. It's a conundrum that we'll have to solve in the spec and outside this PR.

vadi2 commented:
Text updating is fine 👍 

Would you be able to solve the dropdown? It's last thing outstanding:

![image](https://user-images.githubusercontent.com/110988/138546105-885db286-d6e6-43ee-9f2e-4679b9f6da0b.png)

Could you also commit the source for the Mudlet + Discord icon you made to https://github.com/mudlet/artwork?

atari2600tim committed:
redid rectangle with more exact dimensions

atari2600tim commented:
> I see what you mean with [#5490 (comment)](https://github.com/Mudlet/Mudlet/pull/5490#issuecomment-949957251) now. It's a conundrum that we'll have to solve in the spec and outside this PR.

A game developer would rightly not want to waste time and data building and sending rich presence messages to someone who isn't going to use it, so makes sense that it is written like it is.  If they want the button to be available for players who haven't checked the box, then they can send out a GUI that sets it.

I know aiming for deadline, but I will have limited time today and not really sure of details on IRC.

vadi2 commented:
Personally I think it's OK to get it as-is now and fix it up within the next couple of days so we still have a lot of time for testing.

demonnic commented:
I'm fine with taking it as-is and refining the look as we get to release.

atari2600tim commented:
> not really sure of details on IRC.

Is the desire for the IRC option to follow into the other button?
```
defaults, no custom discord invite
generic Discord      (invisible)
-IRC                 -(irrelevent)

custom invite and auto-join channel is still "#mudlet"
TimMUD              Mudlet chat
-(invisible)         -IRC

custom invite and auto-join channel name is not "#mudlet"
TimMUD            Mudlet chat
-IRC              -(invisible)
```

vadi2 commented:
That's a good point about option #3, but I'd keep it in the same spot for consistency.
